Feature/new gene model

andricDu opened the pull request:
Ready for code review. 

Streams in data from Ensembl and NCBI and constructs HashMaps which it then joins onto the correct Nodes (Gene/Transcript/Exon) as they are constructed from the streamed GTF file. Genes are written to mongo as soon as their construction is done. 

NCBI data is ready by the Gene2Xml binary which is downloaded to a temp dir for use. 

Control flow and calculation in the postProcessTranscript function is largely based on the implementation of heliotrope with necessary changes due to the need to do some computation by hand rather than pulling from table columns.

Commit:
Initial commit of new gene model code to replace existing gene import.

Commit:
Added calculations for coding regions and changed field names.

Commit:
Gene importer now uses gene2xml for NCBI data.

Started implementation of Ensembl table joins in memory. 

Still very much prototype code.

Commit:
Protein domains, synonyms, and '+' strand calculations are correct.

TODO: Negative strand calculations need to be corrected.

Commit:
Fixed calculations on negative strand.

Commit:
Adds canonical, name, fixes calculations for lengths and negative strand

Commit:
External db ids without uniprot are joined up.

Commit:
Uniprot ids are now joined up to genes.

Commit:
Dynamic downloading of Gene2Xml tool. Includes check for host platform.

Commit:
Created helper class for parsing files to improve code style.

Commit:
Introduces helper class.

Helper class that takes consumer for parsing lines from stream. This if
for code simplicity and readability. Also DRY.

Commit:
Includes missing BaseReader. Fixes typo in GeneWriter.

Commit:
Fixes transcript 0 cDNA length bug.

Commit:
Bug fix for coding regions and amino acid lengths.

Uses a rounding scheme for amino acid lengths which seems to match
enembl. 

Coding regions fixed for the case where the start and end exons are
actually the same exon.

Commit:
Buffers summary value to fix truncation bug.

Commit:
Comments and some cleanup.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/GeneImporter.java:
```diff
@@ -43,12 +49,12 @@
 
-  public GeneImporter(@NonNull MongoClientURI mongoUri, @NonNull URI genesBsonUri) {
-    this.genesBsonUri = genesBsonUri;
+  public GeneImporter(@NonNull MongoClientURI mongoUri, @NonNull URI gtfUri) {
```
This constructor shouldn't be needed due to `@RequiredArgsConstructor`. Please remove

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/model/ProteinFeature.java:
```diff
@@ -15,41 +15,35 @@
+  private int end;
+
+  public ProteinFeature(@NonNull String interproId, @NonNull String hitName, @NonNull String description) {
+
```
Remove empty lines in constructor

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ASNReader.java:
```diff
@@ -0,0 +1,101 @@
+  @SneakyThrows
+  public Map<String, String> callGene2Xml() {
+
+    Map<String, String> summaryMap = new ConcurrentHashMap<String, String>();
```
`val`

btiernay replied:
Use "diamond operator" on RHS

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ASNReader.java:
```diff
@@ -0,0 +1,101 @@
+    val gzip = new GZIPInputStream(new URL(URI).openStream());
+
+    Process p = Runtime.getRuntime().exec(downloadUtil + " -b T");
+    Thread inThread = new Thread(new ASNInputReader(gzip, p.getOutputStream()));
```
`val`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ASNReader.java:
```diff
@@ -0,0 +1,101 @@
+
+    Process p = Runtime.getRuntime().exec(downloadUtil + " -b T");
+    Thread inThread = new Thread(new ASNInputReader(gzip, p.getOutputStream()));
+    Thread outThread = new Thread(new OutputReader(p.getInputStream(), summaryMap));
```
`val`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ASNReader.java:
```diff
@@ -0,0 +1,101 @@
+    outThread.start();
+    p.waitFor();
+
+    return summaryMap;
```
Wrap in `Collections.unmodifiableMap`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ASNReader.java:
```diff
@@ -0,0 +1,101 @@
+
+    val gzip = new GZIPInputStream(new URL(URI).openStream());
+
+    Process p = Runtime.getRuntime().exec(downloadUtil + " -b T");
```
`val process`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ASNReader.java:
```diff
@@ -0,0 +1,101 @@
+
+    val downloadUtil = getCmd();
+
+    val gzip = new GZIPInputStream(new URL(URI).openStream());
```
`@Cleanup`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ASNReader.java:
```diff
@@ -0,0 +1,101 @@
+  @SneakyThrows
+  public static String getCmd() {
+    val platform = System.getProperty("os.name");
+    log.info("Detecting current platform as: {}", System.getProperty("os.name"));
```
Use `platform`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ASNReader.java:
```diff
@@ -0,0 +1,101 @@
+    } else if (platform.contains("Sun")) {
+      download = "ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools/cmdline/gene2xml.SunOS-5.10-sun4v.gz";
+    } else {
+      throw new RuntimeException("Your platform is not supported for the dcc-import-gene project.");
```
`checkState` and include `platform`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ASNReader.java:
```diff
@@ -0,0 +1,101 @@
+    val gzip = new GZIPInputStream(ftpInputStream);
+
+    val fOut = new FileOutputStream(tmpFile);
+    IOUtils.copy(gzip, fOut);
```
Prefer Guava in general:

http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/io/ByteStreams.html#copy(java.io.InputStream,%20java.io.OutputStream)

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ASNReader.java:
```diff
@@ -0,0 +1,101 @@
+    val ftpInputStream = new URL(download).openStream();
+    val gzip = new GZIPInputStream(ftpInputStream);
+
+    val fOut = new FileOutputStream(tmpFile);
```
Do you need `@Cleanup`s here?

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ASNReader.java:
```diff
@@ -0,0 +1,101 @@
+    val platform = System.getProperty("os.name");
+    log.info("Detecting current platform as: {}", System.getProperty("os.name"));
+
+    String download;
```
Prefer to design out uninitialized variables by extracting a method that uses return statements. `getDownload()` works here. This can determine the platform inside.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/BaseReader.java:
```diff
@@ -0,0 +1,67 @@
+  /**
+   * Constants
+   */
+  private static final Pattern TSV = Pattern.compile("\t");
```
Prefer `Splitters` in `dcc-common-core`. We have one for TSV. In general use Guava.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/BaseReader.java:
```diff
@@ -0,0 +1,67 @@
+   * @param lambda - Lambda encapsulating the callers logic for operating on a line
+   */
+  @SneakyThrows
+  public static void read(String URI, Consumer<String[]> lambda) {
```
`@NonNull`s

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/BaseReader.java:
```diff
@@ -0,0 +1,67 @@
+  public static void read(String URI, Consumer<String[]> lambda) {
+    val bufferedReader = getReader(URI);
+
+    for (String s = bufferedReader.readLine(); null != s; s = bufferedReader.readLine()) {
```
I think you really want Guava's `LineProcessor` here:

http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/io/Resources.html#readLines(java.net.URL,%20java.nio.charset.Charset,%20com.google.common.io.LineProcessor)

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/BaseReader.java:
```diff
@@ -0,0 +1,67 @@
+   */
+  @SneakyThrows
+  public static void read(String URI, Consumer<String[]> lambda) {
+    val bufferedReader = getReader(URI);
```
Don't forget to close these guys with either `try-with-resources` or `@Cleanup`. I prefer the latter because it avoids a nesting level.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/DomainReader.java:
```diff
@@ -0,0 +1,147 @@
+public final class DomainReader {
+
+  private static final String XREF_URI =
+      "ftp://ftp.ensembl.org/pub/grch37/release-82/mysql/homo_sapiens_core_82_37/xref.txt.gz";
```
Factor out a `BASE_URI`.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/DomainReader.java:
```diff
@@ -0,0 +1,147 @@
+      "ftp://ftp.ensembl.org/pub/grch37/release-82/mysql/homo_sapiens_core_82_37/analysis.txt.gz";
+
+  public static Map<String, List<ProteinFeature>> createProteinFeatures(@NonNull Map<String, String> transMap) {
+
```
`DEL`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/DomainReader.java:
```diff
@@ -0,0 +1,147 @@
+            retMap.put(ens, al);
+          }
+        }
+
```
`DEL`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/DomainReader.java:
```diff
@@ -0,0 +1,147 @@
+          if (retMap.containsKey(ens)) {
+            retMap.get(ens).add(pf);
+          } else {
+            List<ProteinFeature> al = new ArrayList<ProteinFeature>();
```
Use `<>` on RHS

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/DomainReader.java:
```diff
@@ -0,0 +1,147 @@
+  public static Map<String, ProteinFeature> interproMap() {
+
+    val descriptionMap = getInterproFromXref();
+    val retMap = new HashMap<String, ProteinFeature>();
```
Use `ImmutableMap.builder` by default always.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/DomainReader.java:
```diff
@@ -0,0 +1,147 @@
+   * @return HashMap of id -> protein feature
+   */
+  public static Map<String, ProteinFeature> interproMap() {
+
```
`DEL`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/DomainReader.java:
```diff
@@ -0,0 +1,147 @@
+   * @return Hashmap of descriptions keyed to Interpro ids.
+   */
+  public static Map<String, String> getInterproFromXref() {
+
```
`DEL`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/DomainReader.java:
```diff
@@ -0,0 +1,147 @@
+    val interproDbId = getInterproDB();
+
+    val retMap = new HashMap<String, String>();
+    BaseReader.read(XREF_URI, line -> {
```
Maybe rename `BaseReader` to `ResourceReader` or `GeneResourceReader`?

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/BaseReader.java:
```diff
@@ -0,0 +1,67 @@
+    for (String s = bufferedReader.readLine(); null != s; s = bufferedReader.readLine()) {
+      s = s.trim();
+      if (s.length() > 0) {
+        String[] line = TSV.split(s);
```
Do you want to trim after split?

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/NameReader.java:
```diff
@@ -0,0 +1,86 @@
+   * Constants
+   */
+  private static final String XREF_URI =
+      "ftp://ftp.ensembl.org/pub/grch37/release-82/mysql/homo_sapiens_core_82_37/xref.txt.gz";
```
`"ftp://ftp.ensembl.org/pub/grch37/release-82/mysql/homo_sapiens_core_82_37"` seems to be repeated a lot. Maybe put it in a shared location.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/NameReader.java:
```diff
@@ -0,0 +1,86 @@
+  /**
+   * Caching
+   */
+  public static Map<String, String> entrezMap = new HashMap<String, String>();
```
Use `<>` on RHS, but prefer `Maps.newHashMap()` or just `newHashMap()`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/SynonymReader.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Class responsible for reading and mapping Gene Synonyms
+ */
+public final class SynonymReader {
```
For these types of classes we typically add `@NoArgConstructor(access = PRIVATE)`. It makes it clear that its a utility class. There is also `@UtilityClass` but it seems buggy so we avoid it in general

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/SynonymReader.java:
```diff
@@ -0,0 +1,59 @@
+  private static final ObjectMapper MAPPER = new ObjectMapper();
+
+  public static Map<String, ArrayNode> getSynonymMap(Map<String, String> idMap) {
+
```
Remove ALL of these EVERYWHERE :)

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/TransReader.java:
```diff
@@ -0,0 +1,124 @@
+   * Constants
+   */
+  private static final String TRANSCRIPT_URI =
+      "ftp://ftp.ensembl.org/pub/grch37/release-82/mysql/homo_sapiens_core_82_37/transcript.txt.gz";
```
Same comment for all URIs. Encapsulate base URI

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
+    ArrayNode transcripts = MAPPER.createArrayNode();
+    ArrayNode exons = MAPPER.createArrayNode();
+
+    for (String s = bufferedReader.readLine(); null != s; s = bufferedReader.readLine()) {
```
This needs to be broken down like so much Jack and the Beanstalk. Please create some helper methods.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
+    val attributes = line[8];
+    val attributeMap = parseAttributes(attributes);
+
+    ObjectNode feature = MAPPER.createObjectNode();
```
`val`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
+    int frame;
+    try {
+      frame = Integer.parseInt(line[7].trim());
+    } catch (Exception var14) {
```
What's `var14`?  Maybe use `ignored` for all of these.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
+
+  private String getName(String symbol) {
+    if (nameMap.containsKey(symbol)) {
+      return nameMap.get(symbol);
```
use `get` with default. Using contains is always a smell for this type of access.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
 
-    log.info("Saving gene documents");
-    saveGenes(genesCollection, genes, geneFilter);
+    String[] tokens = attributes.split(";");
```
`Splitter`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
 
-  private void clearGenes(MongoCollection genesCollection) {
-    genesCollection.drop();
+  private static ObjectNode constructTranscriptNode(ObjectNode data) {
```
Didn't you want to convert this to a pojo?

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
-        genesCollection.save(convertedGene);
+    if (pfs != null) {
+      for (val p : pfs) {
+        val domain = MAPPER.createObjectNode();
```
Pojo?. I notice you have them already. Why not use `putPOJO`? If you are worried about field names, you can change how Jackson serializes them in the configured object mapper. There is an example in the GeneSet imports.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
-      val watch = createStarted();
+    int preExonCdnaEnd = 0;
+    for (JsonNode exon : exons) {
+      ObjectNode exonNode = (ObjectNode) exon;
```
`val`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
 
-      while (genes.hasNextValue()) {
-        val gene = genes.next();
+      exonNode.put("genomic_coding_start", 0);
```
Pojo?

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
-        // Delegate
-        callback.handle(gene);
+    // In case there is no end codon
+    if (endExon.isMissingNode()) {
```
This method needs to be split. Very big! Please add lots of little helpers and try to work bottom up to refactor into a more decomposed design.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
+            val start = cds.get("locationStart").asInt() + 1;
+            transcript.put("coding_region_start", start);
+            exon.put("genomic_coding_start", start);
+            exon.put("cdna_coding_end", exon.get("cdna_coding_start").asInt() + exon.get("genomic_coding_end").asInt()
```
Some comments for things like this would be great. Helper methods will help to break it up and make it easier to read as well. ex. `normalizeXxxx`, `calculateXxxLength`, etc.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
-    log.info("Wrote gene model in {}", watch);
+  private static ObjectNode constructGeneNode(ObjectNode data) {
+    val gene = MAPPER.createObjectNode();
+    gene.put("_gene_id", data.get("gene_id").asText());
```
Do you want `asText` or `textValue` semantics here?

btiernay replied:
If you are doing this a lot in code, I find adding some utilities to do things like `textValue(data, "gene_id")` really helps to cut down on the noise.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
 
-    log.info("Wrote gene model in {}", watch);
+  private static ObjectNode constructGeneNode(ObjectNode data) {
+    val gene = MAPPER.createObjectNode();
```
Pojo?

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
+   * @return Processed Transcript ObjectNode
+   */
+  private ObjectNode postProcessTranscript(ObjectNode transcript, String strand, String canonical) {
+    if (transcript.get("id").asText().equals(canonical)) {
```
``` java
val isCanonical = textValue(transcript,"id").equals(canonical);
transcript.put("is_canonical", isCanonical);
```

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
+      exonNode.put("cdna_coding_end", 0);
+    }
+
+    JsonNode startExon = transcript.path("start_exon");
```
`val` if you can

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
+                exons = MAPPER.createArrayNode();
+              }
+              curTranscript = constructTranscriptNode(entry);
+            } else if (entry.get("type").asText().equals("exon")) {
```
`entry.get("type").asText()` is repeated a lot. Please factor out.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
+          }
+        } else {
+          val entry = parseLine(s);
+          if (entry != null) {
```
Very nested. Try adding some short returns / breaks.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
+      if (s.length() > 0) {
+
+        if (s.charAt(0) == 35) {
+          if (s.startsWith("##fasta")) {
```
Naming variable to give significance to this condition:

``` java
val header = s.startsWith("##fasta");
if (header) { ...
```

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
+      s = s.trim();
+      if (s.length() > 0) {
+
+        if (s.charAt(0) == 35) {
```
Magical! Naming var.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
+  private final Map<String, String> canonicalMap;
+  private final Map<String, List<ProteinFeature>> pFeatures;
+  private final Map<String, ObjectNode> externalIds;
+  private int counter = 0;
```
`State`, not dependencies.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/TransReader.java:
```diff
@@ -0,0 +1,124 @@
+  public static Map<String, String> translationToGene() {
+    val translationMap = getTranslationMap();
+
+    val retMap = new HashMap<String, String>();
```
`Maps.uniqueIndex`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/thread/ASNInputReader.java:
```diff
@@ -15,25 +15,36 @@
+  @SneakyThrows
+  public void run() {
+    int n;
+    byte[] buffer = new byte[16384];
```
Seems pretty low level. Can't we Just use http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/io/ByteStreams.html#copy(java.io.InputStream,%20java.io.OutputStream)?

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,478 @@
+    log.info("CONSUMING GENES");
+    ObjectNode geneNode = null;
+    ObjectNode curTranscript = null;
+    ArrayNode transcripts = MAPPER.createArrayNode();
```
`val`s

btiernay commented on dcc-import-gene/src/test/java/org/icgc/dcc/imports/gene/writer/GeneWriterTest.java:
```diff
@@ -1,127 +0,0 @@
-import com.mongodb.MongoClient;
-import com.mongodb.MongoClientURI;
-
-public class GeneWriterTest {
```
Curious why this was deleted. It can validate the produced JSON docs against a JSON Schema.

btiernay replied:
Bump

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/GeneImporter.java:
```diff
@@ -57,32 +63,50 @@ public ImportSource getSource() {
-    val genesBson = readGenesBson();
+    // TODO: Investigate ordering to see if gene.txt does not need to be read twice.
+    log.info("Doing Ensembl Data Joining...");
+    val idMap = IdReader.getIdMap();
```
I would put all of these values into a model object. Maybe `GeneModel`. This is done elsewhere in the codebase and it makes it easy to pass around and get ordering correct in constructors. I would make it a `@Value` and use a `@Builder` with it.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/GeneImporter.java:
```diff
@@ -57,32 +63,50 @@ public ImportSource getSource() {
     log.info("Writing genes to {}...", mongoUri);
-    writeGenes(genesBson, geneFilter);
+    val writer =
+        new GeneWriter(mongoUri, geneReader, summaryMap, nameMap, synMap, canonicalMap, pfeatures, externalIds);
```
Then the majority of these args will go away

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/GeneImporter.java:
```diff
@@ -57,32 +63,50 @@ public ImportSource getSource() {
-    writeGenes(genesBson, geneFilter);
+    val writer =
+        new GeneWriter(mongoUri, geneReader, summaryMap, nameMap, synMap, canonicalMap, pfeatures, externalIds);
+    writer.consumeGenes();
```
I find it a bit strange that a writer be doing the control flow. To me there should the following actors:
- Importer: Encapsulates the whole import sequence
- Readers: Individual readers for each data source / reference type
- Writer: Writer for the final output stream
- Model(s): Results of the readers and / or the output of the writer
- Processer: Controls the reading and writing process on a per record basis.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/model/ProteinFeature.java:
```diff
@@ -15,41 +15,35 @@
 
-  public static InGeneFilter in(Set<String> geneIds) {
-    return new InGeneFilter(geneIds);
+  public ProteinFeature getCopy() {
```
`copy` would be more idiomatic

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ASNReader.java:
```diff
@@ -0,0 +1,101 @@
+@Slf4j
+public class ASNReader {
+
+  private static final String URI =
```
``` java
/**
 * Constants.
 */
```

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/DomainReader.java:
```diff
@@ -0,0 +1,147 @@
+@Slf4j
+public final class DomainReader {
+
+  private static final String XREF_URI =
```
``` java
/**
 * Constants.
 */
```

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/GeneImporter.java:
```diff
@@ -57,32 +63,50 @@ public ImportSource getSource() {
+    // TODO: Investigate ordering to see if gene.txt does not need to be read twice.
+    log.info("Doing Ensembl Data Joining...");
+    val idMap = IdReader.getIdMap();
+    val nameMap = NameReader.readXrefDisplay();
```
Might want to throw a `val watch = Stopwatch.createStarted();` in here to time each of the legs.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/GeneImporter.java:
```diff
@@ -57,32 +63,50 @@ public ImportSource getSource() {
+    log.info("Doing Ensembl Data Joining...");
+    val idMap = IdReader.getIdMap();
+    val nameMap = NameReader.readXrefDisplay();
+    val synMap = SynonymReader.getSynonymMap(idMap);
```
I would prefix all `*Reader` methods with `read*` for consistency.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/GeneImporter.java:
```diff
@@ -57,32 +63,50 @@ public ImportSource getSource() {
+    log.info("... Done Ensemble Data Joining!");
+
+    log.info("Starting ASN.1 Import from NCBI.");
+    val asnReader = new ASNReader();
```
`ASNReader` speaks more to the serialization format than to the content. Can we rename to something like `GTFReader`?

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ExternalReader.java:
```diff
@@ -0,0 +1,119 @@
+        String xrefId = line[3];
+        if (NameReader.uniprotMap.containsKey(xrefId)) {
+          ArrayNode arrayNode = (ArrayNode) externalDbs.get("uniprotkb_swissprot");
+          String uniprot = NameReader.uniprotMap.get(xrefId);
```
We should avoid using statics like this as it makes testing more difficult because there is an implicit ordering that isn't enforced by a constructor. My opinion is that this class should have its dependencies passed to it.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/GeneReader.java:
```diff
@@ -0,0 +1,86 @@
+   * @return Map of stable_id (ENSG*) -> CanonicalTranscript
+   */
+  @SneakyThrows
+  public static Map<String, String> canonicalMap() {
```
`readCanonicalMap`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/NameReader.java:
```diff
@@ -0,0 +1,86 @@
+    val retMap = new HashMap<String, String>();
+    BaseReader.read(XREF_URI, line -> {
+      // Only use the rows we care about.
+      if ("12600".equals(line[1])) {
```
``` java
val source = line[1];
if (isGeneWiki(source))
...
```

Code is always better than comments and naming variables / helper methods aid a lot in that regard.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/SynonymReader.java:
```diff
@@ -0,0 +1,59 @@
+      "ftp://ftp.ensembl.org/pub/grch37/release-82/mysql/homo_sapiens_core_82_37/external_synonym.txt.gz";
+  private static final ObjectMapper MAPPER = new ObjectMapper();
+
+  public static Map<String, ArrayNode> getSynonymMap(Map<String, String> idMap) {
```
`readSynonymMap`

Commit:
Fix for OBO on negative strand, started PR feedback and refactoring.

Commit:
Completed decomposition of readers. Next step is dependency injection.

TODO: Encapsulation of readers inside a model class, dependency
injection, refactoring of gene constructor.

Commit:
Ensembl data is now encapsulated in a class with a builder.

Commit:
Continued refactor. 

Cleanup of gene constructor class.
Precompiled regex before use. 
Using splitters from dcc-commons. 
Default description is now empty string rather than null.

Commit:
Additional commenting, break up consumeGenes and misc refactoring

andricDu commented:
Probably good for another review. 

Still not ready for merge pending final review from @junjun-zhang and pruning of CDS object from coding exons.

andricDu commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,349 @@
+  }
+
+  /**
+   * Helper for setting the final transcript values
```
TODO: Fix this comment

andricDu commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,349 @@
+   * @return Processed Transcript ObjectNode
+   */
+  private ObjectNode postProcessTranscript(ObjectNode transcript, String strand, String canonical) {
+    if (transcript.get("id").asText().equals(canonical)) {
```
`asText()` helper

andricDu commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/thread/ASNInputReader.java:
```diff
@@ -15,21 +15,34 @@
+  @Override
+  @SneakyThrows
+  public void run() {
+    ByteStreams.copy(in, out);
```
Does not appear to have impacted performance.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/GeneImporter.java:
```diff
@@ -42,47 +42,48 @@
   }
 
+  /**
+   * Main pipeline execution for creating gene model. Calls all readers to pre-compute hashmaps of relevant information
```
:+1:

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,349 @@
+   * @return ObjectNode representation of the gtf row.
+   */
+  private ObjectNode parseLine(@NonNull String s) {
+    String[] line = TAB.splitToList(s).stream().toArray(String[]::new);
```
Simplify as before. Also you can specify trim on the `Splitter`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,349 @@
+    val attributeMap = parseAttributes(attributes);
+
+    int strandNumber = 0;
+    if (strand == '+') {
```
Don't compare strings with equals! Did you run findbugs locally?

btiernay replied:
Nvm, its a char!

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,349 @@
+    }
+
+    val negative = locationStart <= 0 && locationEnd <= 0;
+    assert strand == '-' == negative;
```
Use `checkState`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,349 @@
+      strandNumber = 1;
+    } else if (strand == '-') {
+      strandNumber = -1;
+    }
```
Replace with:

`val strandNumber = convertStand(strand)`

``` java
private static int convertStand(char strand){
if (strand == '+') {
  return 1
} else if (strand == '-') {
 return -1;
} else {
 return 0;
}
```

This reduces noise in the mainline code making the overall logic more apparent (less noise).

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,349 @@
+
+  /**
+   * This method performs the calculations on transcript and exons to correctly mark coding regions. Heavily based on:
+   * https://github.com/icgc-dcc/dcc-heliotrope/blob/working/src/main/scripts/Heliotrope/Update/Ensembl.pm#L885
```
Add newline between description and `@params` section.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ExternalReader.java:
```diff
@@ -0,0 +1,127 @@
+  @Getter
+  private final Map<String, ObjectNode> externalIds = new HashMap<>();
+
+  public ExternalReader read() {
```
Still a bit long. As a general rule, a method body should fit on a screen. Any way to extract some helpers?

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ExternalReader.java:
```diff
@@ -0,0 +1,127 @@
+          ArrayNode arrayNode = (ArrayNode) externalDbs.get("uniprotkb_swissprot");
+          String uniprot = nameReader.getUniprotMap().get(xrefId);
+
+          Iterator<JsonNode> iter = arrayNode.elements();
```
Like here: `hasUniprot(arrayNode)`? Actually, I think `ArrayNode` has methods for this.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/TranscriptProcessing.java:
```diff
@@ -0,0 +1,234 @@
+   */
+  public static ArrayNode exonDefaults(ArrayNode exons) {
+    int preExonCdnaEnd = 0;
+    for (JsonNode exon : exons) {
```
`val`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/TranscriptProcessing.java:
```diff
@@ -0,0 +1,234 @@
+   * @param exon ObjectNode representation of an exon.
+   * @param strand Flag which determines if we are working on a positive or negative strand.
+   */
+  public static void computeStartRegion(ObjectNode transcript, ObjectNode exon, String strand) {
```
`@NonNull`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/InterproReader.java:
```diff
@@ -0,0 +1,50 @@
+@RequiredArgsConstructor
+public class InterproReader {
+
+  private final XrefReader xrefReader;
```
`@NonNull`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/DomainReader.java:
```diff
@@ -0,0 +1,71 @@
+    val retMap = new HashMap<String, List<ProteinFeature>>();
+
+    BaseReader.read(PROTEIN_FEATURE_URI, line -> {
+      if (analysisMap.containsKey(line[7])) {
```
Short returns for both of these nested blocks

Commit:
Further PR feedback work.

btiernay commented:
I'm getting this when I try to build:

```
Description Resource    Path    Location    Type
The constructor GeneImporter(MongoClientURI, URI) is undefined  Importer.java   /dcc-import-client/src/main/java/org/icgc/dcc/imports/client/core   line 125    Java Problem
The constructor GeneImporter(MongoClientURI, URI) is undefined  PathwayImporterTest.java    /dcc-import-pathway/src/test/java/org/icgc/dcc/imports/pathway  line 39 Java Problem
```

btiernay commented:
ok to test

btiernay commented:
Findbugs reports these issues:

```
Description Resource    Path    Location    Type
Immediate dereference of the result of readLine() in org.icgc.dcc.imports.gene.core.GeneConstructor.consumeGenes() [Of Concern(15), Normal confidence]  GeneConstructor.java    /dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core   line 86 FindBugs Problem (Of concern)
Immediate dereference of the result of readLine() in org.icgc.dcc.imports.gene.reader.BaseReader.read(String, Consumer) [Of Concern(15), Normal confidence] BaseReader.java /dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader line 52 FindBugs Problem (Of concern)
```

```
Bug: Immediate dereference of the result of readLine() in org.icgc.dcc.imports.gene.reader.BaseReader.read(String, Consumer)
The result of invoking readLine() is immediately dereferenced. If there are no more lines of text to read, readLine() will return null and dereferencing that will generate a null pointer exception.

Rank: Of Concern (15), confidence: Normal
Pattern: NP_IMMEDIATE_DEREFERENCE_OF_READLINE 
Type: NP, Category: STYLE (Dodgy code)
```

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/BaseReader.java:
```diff
@@ -0,0 +1,69 @@
+ * Helper class for reading files and consuming lines.
+ */
+@Slf4j
+public final class BaseReader {
```
I would rename this to `TSVReader`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/BaseReader.java:
```diff
@@ -0,0 +1,69 @@
+    val watch = Stopwatch.createStarted();
+    val bufferedReader = getReader(URI);
+
+    for (String s = bufferedReader.readLine().trim(); null != s; s = bufferedReader.readLine()) {
```
Probably cleaner to do something like this: http://stackoverflow.com/a/10811303/527333

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/AnalysisReader.java:
```diff
@@ -0,0 +1,47 @@
+  private final Map<String, String> analysisMap = new HashMap<>();
+
+  public AnalysisReader read() {
+    BaseReader.read(ANALYSIS_URI, line -> {
```
I would externalize `ANALYSIS_URI`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/AnalysisReader.java:
```diff
@@ -0,0 +1,47 @@
+    return this;
+  }
+
+}
```
Consider this alternative design for the readers:

# Base Class

``` java
@Slf4j
@RequiredArgsConstructor
public abstract class TsvReader {

  /**
   * Configuration
   */
  @NonNull
  private final String uri;

  @SneakyThrows
  protected Stream<List<String>> readRecords() {
    log.info("Reading '{}'...", uri);

    return reader().lines().map(TAB::splitToList);
  }

  private BufferedReader reader() throws MalformedURLException, IOException {
    val gzip = new GZIPInputStream(new URL(uri).openStream());
    val reader = new InputStreamReader(gzip, UTF_8);

    return new BufferedReader(reader);
  }

}
```

# Example Reader

``` java
public class AnalysisReader2 extends TsvReader {

  public AnalysisReader2(String uri) {
    super(uri);
  }

  public Map<String, String> read() {
    return readRecords()
        .filter(this::isPfam)
        .collect(toMap(this::getId, this::getGffSource));
  }

  private boolean isPfam(List<String> record) {
    return "pfam".equals(getGffSource(record));
  }

  private String getId(List<String> record) {
    return record.get(0);
  }

  private String getGffSource(List<String> record) {
    return record.get(6);
  }

}
```

This is a more functional design that leverages the Java 8 streams API. Parsing is handled to the base class. The `Stream<List<String>>` abstraction allows powerful streaming capabilities, including filtering, mapping and collecting. You could take this a step further and create other base classes that handle certain data types (Map<String String>) and only require a developer to specify a key / value mapper. In the end, this design gives the implementer flexibility and statelessness, with a minimal amount of code.

Commit:
Fixes build by updating call to gene constructor in pathway and client.

Commit:
Refactor towards using streams.

Updated readers, encapsulated mappings into model classes.

Commit:
ConsumeGenes now uses Streams API.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/Sources.java:
```diff
@@ -0,0 +1,36 @@
+ */
+package org.icgc.dcc.imports.gene.core;
+
+public final class Sources {
```
`@NoArgsConstructor(access = PRIVATE)`. Prevents instantiation, but more importantly signals that this is a utility class. My preference would be to use `@Utility` as it removes some boilerplate, but it can cause issue with `static` imports.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/TransJoiner.java:
```diff
@@ -0,0 +1,50 @@
+    for (val entry : translationMap.entrySet()) {
+      retMap.put(entry.getValue(), transcriptMap.get(entry.getKey()));
+    }
+    return retMap;
```
Not sure this is more readable, but it does at least make it immutable :)

``` java
    return translationMap.entrySet().stream()
        .collect(toImmutableMap(e -> e.getValue(), e -> transcriptMap.get(e.getKey())));
```

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/TransJoiner.java:
```diff
@@ -0,0 +1,50 @@
+@RequiredArgsConstructor
+public class TransJoiner {
+
+  private final TranslationMapping translationReader;
```
`@NonNull` x 2

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/TranscriptProcessing.java:
```diff
@@ -0,0 +1,238 @@
+  public static ArrayNode exonDefaults(ArrayNode exons) {
+    int preExonCdnaEnd = 0;
+    for (val exon : exons) {
+      ObjectNode exonNode = (ObjectNode) exon;
```
`val`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/TranscriptProcessing.java:
```diff
@@ -0,0 +1,238 @@
+      exonNode.put("genomic_coding_end", 0);
+      exonNode.put("cdna_coding_start", 0);
+      exonNode.put("cdna_coding_end", 0);
+    }
```
`\n`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/TranscriptProcessing.java:
```diff
@@ -0,0 +1,238 @@
+
+  /**
+   * Helper for computing the values of the beginning of the coding region of a transcript and start exon, both with the
+   * coordinate system of the DNA and the coordinate system of the transcript sequence (cDNA).
```
`\n`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/TranscriptProcessing.java:
```diff
@@ -0,0 +1,238 @@
+  }
+
+  /**
+   * Helper for computing the values of the beginning of the coding region of a transcript and start exon, both with the
```
comment == :+1:

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/TranscriptProcessing.java:
```diff
@@ -0,0 +1,238 @@
+      transcript.put("coding_region_start", cdsStart);
+      exon.put("genomic_coding_start", cdsStart);
+      exon.put("cdna_coding_start", asInt(exon, "cdna_start") + (cdsStart - asInt(exon, "start") + 1));
+    }
```
`\n` almost always after closing brace. Wish Eclipse formatter was more expressive...

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/TranscriptProcessing.java:
```diff
@@ -0,0 +1,238 @@
+   * @param data ObjectNode holding raw json data
+   * @return ObjectNode representing a transcript with default values.
+   */
+  public static ObjectNode constructTranscriptNode(ObjectNode data) {
```
`@NonNull`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/TranscriptProcessing.java:
```diff
@@ -0,0 +1,238 @@
+/**
+ * Collection of stateless helpers used for processing transcripts and exons
+ */
+public final class TranscriptProcessing {
```
`NoArg`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/model/Ensembl.java:
```diff
@@ -0,0 +1,39 @@
+@Builder
+public class Ensembl {
+
+  private Map<String, String> nameMap;
```
`private` not required, but I will not argue if you are trying to make a point :)

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/model/GeneMapping.java:
```diff
@@ -15,25 +15,23 @@
 
+  Map<String, String> geneIdMap;
+  Map<String, String> canonicalMap;
+  Multimap<String, String> xrefGeneMap;
```
`\n`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/model/ProteinFeature.java:
```diff
@@ -15,41 +15,33 @@
 
-  public static InGeneFilter in(Set<String> geneIds) {
-    return new InGeneFilter(geneIds);
+  public ProteinFeature copy() {
```
Checkout `@Wither`. Works nice with `@Value` to make "streamable" immutable objects.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/model/TranslationMapping.java:
```diff
@@ -15,21 +15,20 @@
+public class TranslationMapping {
 
+  Map<String, String> translationMap;
+  Map<String, String> translationToGene;
```
`\n` before every class closing. rationale: symmetry with class opening.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ASNReader.java:
```diff
@@ -0,0 +1,107 @@
+    } else if (platform.contains("Sun")) {
+      return BASE_URI + "gene2xml.SunOS-5.10-sun4v.gz";
+    } else {
+      throw new RuntimeException("Your platform is not supported for the dcc-import-gene project.");
```
Nice

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/AnalysisReader.java:
```diff
@@ -0,0 +1,49 @@
+        .collect(toMap(this::getId, this::getGffSource));
+  }
+
+  private boolean isPfam(List<String> record) {
```
Ideally these would be `static`, but it makes things more noisy on the ground floor, so I think this is better here :)

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/DomainReader.java:
```diff
@@ -0,0 +1,109 @@
+          }
+        }
+      }
+    });
```
I think we can stream this bad boy nicely... at least most of it.

``` java
readRecords().filter(this::hasAnalysis).filter(this::hasInterpo).map(this::createProteinFeature).forEach(...)`
```

Btw, you might want to consider a `MultiMap` for this. Would autovivify your lists for you instead of having to manually create them. Otherwise you can do:

``` java
retMap.computeIfAbsent(ens, ArrayList::new).add(pf);
```

andricDu replied:
cool. Yea I like the `Multimap` idea, because then createProteinFeature can return an `Entry` and the whole thing becomes:

```
readRecords().filter(this::hasInterproAnalysis).map(this::createProteinFeature).forEach(retMap::put);
```

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/DomainReader.java:
```diff
@@ -0,0 +1,109 @@
+    return ImmutableMap.<String, List<ProteinFeature>> copyOf(retMap);
+  }
+
+  private String getAnalysisId(List<String> record) {
```
These should be `static`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ExternalDatabaseReader.java:
```diff
@@ -0,0 +1,51 @@
+        .filter(this::isInterpro)
+        .map(this::getDBId)
+        .collect(toList())
+        .get(0);
```
No need to create the whole list to get the first:

``` java
    return readRecords()
        .filter(record -> record.size() > 1)
        .filter(this::isInterpro)
        .map(this::getDBId)
        .findFirst().orElse(null);
```

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ExternalReader.java:
```diff
@@ -0,0 +1,150 @@
+      }
+    });
+
+    return ImmutableMap.<String, ObjectNode> copyOf(externalIds);
```
`ImmutableMap.copyOf(externalIds);` should work due to return type filling in the details.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ExternalReader.java:
```diff
@@ -0,0 +1,150 @@
+    this.xrefMapping = xrefMapping;
+    this.getIdMap = geneMapping.getGeneIdMap();
+    this.translationToGene = translationMapping.getTranslationToGene();
+
```
Remove

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ExternalReader.java:
```diff
@@ -0,0 +1,150 @@
+
+  private void handleUniprotIds(List<String> record, Map<String, ObjectNode> externalIds) {
+    // Uniprot Ids are for proteins, which means we match them to translations and eventually work up to a gene/
+    String geneId = getIdMap.get(translationToGene.get(getItemId(record)));
```
More `val` bell!

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/InterproReader.java:
```diff
@@ -0,0 +1,69 @@
+   */
+  public Map<String, ProteinFeature> read() {
+
+    val interproMap = ImmutableMap.<String, ProteinFeature> builder();
```
Stream:

``` java
readRecords().map(this::createProteinFeature).collect(toImmutableMap(ProteinFeature::getHitName));
```

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/InterproReader.java:
```diff
@@ -0,0 +1,69 @@
+   * Returns a map of protein features
+   */
+  public Map<String, ProteinFeature> read() {
+
```
Remove `\n`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/InterproReader.java:
```diff
@@ -0,0 +1,69 @@
+    return interproMap.build();
+  }
+
+  private String getInterproId(List<String> record) {
```
`static`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/SynonymReader.java:
```diff
@@ -0,0 +1,76 @@
+          }
+        }
+      }
+    });
```
``` java
  public Map<String, ArrayNode> read() {
    val synMap = new HashMap<String, ArrayNode>();
    readRecords().forEach(record -> {
      for (String id : getExternalIds(record)) {
        ArrayNode synonyms = synMap.computeIfAbsent(id, (x) -> DEFAULT.createArrayNode());
        synonyms.add(getSynonym(record));
      }
    });

    return synMap;
  }
```

or

``` java
  public Map<String, ArrayNode> read() {
    val synMap = new HashMap<String, ArrayNode>();
    readRecords().forEach(record -> {
      getExternalIds(record).forEach(id -> {
        ArrayNode synonyms = synMap.computeIfAbsent(id, (x) -> DEFAULT.createArrayNode());
        synonyms.add(getSynonym(record));
      });
    });

    return synMap;
  }
```

or the slightly illegible but terse:

``` java
    readRecords()
        .forEach(r -> getExternalIds(r)
            .forEach(id -> synMap.computeIfAbsent(id, x -> DEFAULT.createArrayNode()).add(getSynonym(r))));
```

or slightly more typey:

``` java
    Function<String, ArrayNode> newArrayNode = ignore -> DEFAULT.createArrayNode();
    readRecords()
        .forEach(r -> getExternalIds(r)
            .forEach(id -> synMap.computeIfAbsent(id, newArrayNode))
               .add(getSynonym(r))));
```

In all of these, just make sure `getExternalIds` returns a list always.

btiernay replied:
Actually, now that I see how the code is being used, I too think that a multimap may be appropriate. See http://stackoverflow.com/questions/23003542/cleanest-way-to-create-a-guava-multimap-from-a-java8-stream. In particular, `MultimapCollector` and/or `groupingBy`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/TranscriptReader.java:
```diff
@@ -0,0 +1,60 @@
+    readRecords().forEach(record -> {
+      transcriptMapBuilder.put(getId(record), getStableId(record));
+      transcriptToGeneBuilder.put(getId(record), getGeneId(record));
+    });
```
Would it simplify things to just return a single `Map` here?

``` java
@Value public class Transcript {
  @NonNull String id;
  @NonNull String stableId;
  @NonNull String geneId;
}
```

``` java
return readRecords().collect(toImmutableMap(this:: getId, this::createTranscript));
```

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/XMLHandler.java:
```diff
@@ -0,0 +1,96 @@
+      summaryMap.put(currentId, currentValue.toString());
+    }
+    this.elementStack.pop();
+
```
Remove `\n`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/XMLHandler.java:
```diff
@@ -0,0 +1,96 @@
+      currentValue = new StringBuilder();
+    }
+    this.elementStack.push(qName);
+
```
Remove `\n`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/XMLHandler.java:
```diff
@@ -0,0 +1,96 @@
+  public void startElement(String uri, String localName, String qName, Attributes attributes) {
+    if ("Entrezgene_summary".equals(qName)) {
+      currentValue = new StringBuilder();
+    }
```
Quick, what's the rule after almost all braces?

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/XrefReader.java:
```diff
@@ -0,0 +1,124 @@
+   */
+  public XrefMapping read() {
+    // HashMap needed for name/entrez due to duplicate entries for some uncharacterized locations.
+    val nameMapBuilder = new HashMap<String, String>();
```
Very Nice!

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/XrefReader.java:
```diff
@@ -0,0 +1,124 @@
+        .build();
+  }
+
+  private boolean isGeneWiki(List<String> record) {
```
`static` as a cue to inform the reader this does not modify the class.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/thread/ASNInputReader.java:
```diff
@@ -0,0 +1,48 @@
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class ASNInputReader implements Runnable {
```
`@RequiredArgsConstructor`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/thread/ASNInputReader.java:
```diff
@@ -0,0 +1,48 @@
+@Slf4j
+public class ASNInputReader implements Runnable {
+
+  private InputStream in;
```
`@NonNull` / `final`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/thread/OutputReader.java:
```diff
@@ -0,0 +1,54 @@
+import lombok.SneakyThrows;
+import lombok.val;
+
+public class OutputReader implements Runnable {
```
Same

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,44 @@
+   * State
+   */
+  private int counter = 0;
+  private MongoCollection geneCollection;
```
`final`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,44 @@
+  /**
+   * State
+   */
+  private int counter = 0;
```
Remove and return from `writeFiles`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,44 @@
-          formatCount(insertCount), formatCount(excludeCount));
-    } finally {
-      genes.close();
+  public void writeFiles(ObjectNode value) {
```
Rename to `writeGenes`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,351 @@
+  @SneakyThrows
+  public void consumeGenes() {
+    log.info("CONSUMING GENES");
+    bufferedReader.lines().filter(this::isData).map(this::parseLine).forEach(this::handleEntry);
```
Isn't this just a `TsvReader` as well?

andricDu replied:
Good point!

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,351 @@
+  /**
+   * State
+   */
+  ObjectNode geneNode = null;
```
Ideally this would be projected so the class could be stateless

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,351 @@
+   * Dependencies
+   */
+  private final BufferedReader bufferedReader;
+  private final Map<String, String> summaryMap;
```
`@NonNull`s as appropriate.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,351 @@
+  }
+
+  private void writeGene(@NonNull ObjectNode value) {
+    writer.writeFiles(value);
```
`writer.writeGene`

andricDu replied:
writeFile is a required method signature due to the Jongo abstract class. Want me to alias it with a `writeGene` method?

btiernay replied:
You sure? I didn't see an `@Override` on it.

btiernay replied:
You're right. We should rename that method to `write` in `AbstractJongoWriter`. Thanks

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,351 @@
+   * @return ObjectNode representation of the gtf row.
+   */
+  private ObjectNode parseLine(@NonNull String s) {
+    String[] line = TAB.trimResults().splitToList(s).stream().toArray(String[]::new);
```
`trimResults()` creates a new `Splitter` each time. Maybe cache this expression in a static field.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,351 @@
+
+  /**
+   * Responsible for parsing the lines of the gtf file and producing generic ObjectNodes
+   * 
```
:+1: for the `\n`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,351 @@
+      transcript.put("is_canonical", true);
+    } else {
+      transcript.put("is_canonical", false);
+    }
```
`transcript.put("is_canonical", isCanonical(transcript));` would reduce 4 lines from this already long-ish mainline.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,351 @@
+        computeStartRegion(transcript, exon, strand);
+      }
+
+      // End Exon. Note: This can be the same exon as the start exon.
```
Good comments. Thanks.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,351 @@
+    val attributeMap = new HashMap<String, String>();
+    val tokens = SEMICOLON.split(attributes);
+    for (val token : tokens) {
+      String[] kv = UNKNOWN_WHITESPACE.split(QUOTES.matcher(token.trim()).replaceAll(""));
```
You'll have to explain what this is doing as I'm not exactly sure.

andricDu replied:
Yea this took me a second. 

Is this more clear?

``` java
    return SEMICOLON.splitToList(keyValueAttributes).stream()
        .map(GeneConstructor::stripQuotes)
        .map(UNKNOWN_WHITESPACE::split)
        .filter(GeneConstructor::isValidAttribute)
        .collect(Collectors.toMap(attr -> attr[0], attr -> attr[1]));
```

btiernay replied:
Yes, much! Are you still trimming though? Needed?

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/GeneImporter.java:
```diff
@@ -42,47 +42,48 @@
-  @NonNull
-  private final URI genesBsonUri;
-  @NonNull
+  private final URI gtfUri;
```
`@NonNull` as appropriate.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/GeneImporter.java:
```diff
@@ -42,47 +42,48 @@
-    writeGenes(genesBson, geneFilter);
+    log.info("Starting ASN.1 Import from NCBI.");
+    val asnReader = new ASNReader();
+    val summaryMap = asnReader.callGene2Xml();
```
`callGene2Xml` -> `readSummary` ?

Commit:
Style, non nulls, streams and find bugs.

Commit:
Updateds dcc-commons version, find bug fixes, more refactoring, fix OBO

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,371 @@
+    val attributes = line[8];
+    val attributeMap = parseAttributes(attributes);
+
+    ObjectNode feature = DEFAULT.createObjectNode();
```
`val`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,371 @@
+    feature.put("strand", strandNumber);
+    for (val kv : attributeMap.entrySet()) {
+      feature.put(kv.getKey(), kv.getValue());
+    }
```
`\n`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,371 @@
+    return feature;
+  }
+
+  private int convertStrand(char strand) {
```
`static`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java:
```diff
@@ -0,0 +1,371 @@
+   * @return A Map of attribute name to attribute value.
+   */
+  private static Map<String, String> parseAttributes(String keyValueAttributes) {
+    return SEMICOLON.splitToList(keyValueAttributes).stream()
```
:+1:

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/Sources.java:
```diff
@@ -0,0 +1,41 @@
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor(access = PRIVATE)
+public final class Sources {
```
Thanks!

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/TranscriptProcessing.java:
```diff
@@ -0,0 +1,239 @@
+    transcript.putNull("length");
+    transcript.putNull("length_amino_acid");
+    transcript.putNull("length_cds");
+    transcript.put("domains", DEFAULT.createArrayNode());
```
`\n`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/thread/OutputReader.java:
```diff
@@ -0,0 +1,54 @@
+    factory.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
+    val handler = new XMLHandler(summaryMap);
+
+    SAXParser saxParser = factory.newSAXParser();
```
`val`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/thread/OutputReader.java:
```diff
@@ -0,0 +1,54 @@
+  @SneakyThrows
+  public void run() {
+    val factory = SAXParserFactory.newInstance();
+    factory.setValidating(false);
```
Are these 3 lines required? Just curious.

andricDu replied:
Unfortunately yes, otherwise it tries to load the dtd referenced in the xml, tries to validate that and validate the xml itself, none of which are technically valid.  Also the external dtd does not actually exist.

btiernay replied:
I would add that exact thing as a comment.

btiernay commented on dcc-import-pathway/src/main/java/org/icgc/dcc/imports/pathway/reader/PathwayUniprotReader.java:
```diff
@@ -38,14 +38,10 @@
-      FieldNames.PATHWAY_SPECIES
-  };
+  private static final String[] CSV_HEADER =
+      { FieldNames.PATHWAY_UNIPROT_ID, FieldNames.PATHWAY_REACTOME_ID, FieldNames.PATHWAY_URL, // Not used
```
Maybe restore the previous formatting if possible.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/XrefReader.java:
```diff
@@ -0,0 +1,124 @@
+        .build();
+  }
+
+  private boolean isInterpro(List<String> record) {
```
`static`s

andricDu replied:
Relies on a instance field.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/XMLHandler.java:
```diff
@@ -0,0 +1,96 @@
+
+  public List<String> ids = new ArrayList<String>();
+
+  public XMLHandler(Map<String, String> summaryMap) {
```
`@RequiredArgsConstructor` / `@NonNull`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/XMLHandler.java:
```diff
@@ -0,0 +1,96 @@
+  private StringBuilder currentValue;
+  public Map<String, String> summaryMap;
+
+  public List<String> ids = new ArrayList<String>();
```
`final`s where possible

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/XMLHandler.java:
```diff
@@ -0,0 +1,96 @@
+    String value = new String(ch, start, length).trim();
+    if (value.length() == 0) return;
+
+    if (isEnsembl && isGene() && currentElement().equals("Object-id_str")) {
```
`val <condition | state> = isEnsembl && isGene() && currentElement().equals("Object-id_str")`

or

`if (is<condition | state>()){`

Both reveal a rationale, state transition, etc. by naming

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/XrefReader.java:
```diff
@@ -0,0 +1,124 @@
+    return record.get(0);
+  }
+
+  private String getDbId(List<String> record) {
```
`static`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java:
```diff
@@ -17,139 +17,44 @@
+  private int counter = 0;
+  private final MongoCollection geneCollection;
+
+  /**
```
No need for such comments. Better off not having them.

btiernay commented on dcc-import-diagram/src/main/java/org/icgc/dcc/imports/diagram/reader/DiagramXmlReader.java:
```diff
@@ -23,37 +23,20 @@
+  @SneakyThrows
+  public String readPathwayXml(@NonNull String dbId) {
     val url = new URL(format(DIAGRAM_XML_URL, dbId));
-    return escape(Resources.toString(url, UTF_8));
```
Thanks!

Commit:
Further PR feedback.

Commit:
Moved to fully streaming architecture. 

Implemented joiners and processors for operations on streamed elements. 
Renamed jongo writer method. 

Todo: GeneWrite takes stream and unit testing.

Commit:
Fixes OBO on cdna_coding_start, fixes end_exon=0 edge case, adds phase.

PR feedback.

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/joiner/EnsemblJoiner.java:
```diff
@@ -70,6 +74,19 @@ private void attachProteinFeatures(ObjectNode transcript) {
+        exonNode.remove("id");
+      }
+    }
+
```
Remove `\n`

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ExonReader.java:
```diff
@@ -0,0 +1,50 @@
+  }
+
+  public Map<String, Entry<Integer, Integer>> read() {
+    return readRecords().collect(Collectors.toImmutableMap(r -> getStableId(r), r -> getPhaseTuple(r)));
```
Static import

btiernay replied:
``` java
return readRecords().collect(toImmutableMap(this::getStableId, this::getPhaseTuple));
```

btiernay commented on dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/joiner/EnsemblJoiner.java:
```diff
@@ -45,7 +46,10 @@ public ObjectNode join(ObjectNode gene) {
     gene.put("external_db_ids", externalIds().get(asText(gene, "_gene_id")));
```
``` java
 @NonNull
 private final Ensembl ensembl;
```

Commit:
First pass at some unit testing. Additional PR feedback.

Better logic for isData conditional.

Commit:
Reimplemented GeneWriterTest and fixed validation errors from that test

Commit:
Adds schema file.
